orttraining/orttraining/training_api/checkpoint.cc

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(ENABLE_TRAINING) && defined(ENABLE_TRAINING_ON_DEVICE)
+
+#include "core/common/logging/logging.h"
+#include "core/common/logging/sinks/clog_sink.h"
+#include "core/platform/path_lib.h"
+#include "core/platform/env.h"
+#include "orttraining/core/framework/checkpointing.h"
+#include "orttraining/training_api/checkpoint.h"
+#include <type_traits>
+#include "core/util/protobuf_parsing_utils.h"
+#include "orttraining/core/framework/protobuf_message_sequence.h"
+#include "onnx/defs/tensor_proto_util.h"
+#include "core/framework/tensorprotoutils.h"
+
+namespace onnxruntime {
+namespace training {
+namespace api_test {
+
+PathString CreateFolderIfNotExists(const PathString& path, const std::string& folder_name) {
+  PathString new_folder_path = path + GetPathSep<PathChar>() + ORT_TSTR(folder_name);
+  LOGS_DEFAULT_IF(Env::Default().FolderExists(new_folder_path), WARNING)
+      << ToUTF8String(new_folder_path) << " directory exists - data may be overwritten.";
+
+  ORT_ENFORCE(Env::Default().CreateFolder(new_folder_path).IsOK());
+
+  return new_folder_path;
+}
+
+Status CreateOrtValuesFromTensorProtos(
+    const std::vector<ONNX_NAMESPACE::TensorProto>& tensor_protos,
+    NameMLValMap& name_to_ort_value) {
+  static const OrtMemoryInfo cpu_alloc_info{onnxruntime::CPU, OrtDeviceAllocator};
+  std::vector<std::vector<char>> tensor_buffers{};
+
+  for (const auto& tensor_proto : tensor_protos) {
+    const auto* tensor_type = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto.data_type());
+    const size_t element_size = tensor_type->GetElementType()->Size();
+    const TensorShape shape{
+        tensor_proto.dims().data(), static_cast<size_t>(tensor_proto.dims().size())};
+
+    std::vector<char> tensor_buffer{};
+    tensor_buffer.resize(element_size * shape.Size());
+
+    const MemBuffer mem_buffer{tensor_buffer.data(), tensor_buffer.size(), cpu_alloc_info};
+
+    OrtValue ort_value;
+
+    ORT_RETURN_IF_ERROR(onnxruntime::utils::TensorProtoToMLValue(
+        Env::Default(), nullptr, tensor_proto, mem_buffer,
+        ort_value));
+
+    name_to_ort_value.emplace(tensor_proto.name(), ort_value);
+    tensor_buffers.emplace_back(std::move(tensor_buffer));
+  }
+
+  return Status::OK();
+}
+
+const std::vector<std::string> ParseStringData(
```

Comment:
we have utilities in both ort and onnx for parsing data from tensorproto, we should just leverage them

Reply:
Yeah, I added the std::string type support because the onnx utility did not provide this type support. :)

Reply:
Please check UnpackTensor : https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/framework/tensorprotoutils.h#L417